src/ray/core_worker/object_interface.cc

```diff
@@ -80,24 +94,107 @@ Status CoreWorkerObjectInterface::Get(
+    } else {
+      iter->second++;
+    }
+  }
+
+  auto current_timeout_ms = timeout_ms;
+
+  for (const auto &entry : object_ids_per_store_provider) {
+    std::unordered_set<ObjectID> objects;
+    auto start_time = current_time_ms();
+    RAY_RETURN_NOT_OK(
+        Wait(entry.first, entry.second, num_objects, current_timeout_ms, &objects));
+    int64_t duration = current_time_ms() - start_time;
+    current_timeout_ms =
+        (current_timeout_ms == -1)
+            ? current_timeout_ms
+            : std::max(static_cast<int64_t>(0), current_timeout_ms - duration);
+    for (const auto &entry : objects) {
+      num_objects -= object_counts[entry];
+    }
+
+    for (size_t i = 0; i < ids.size(); i++) {
+      if (objects.count(ids[i]) > 0) {
+        (*results)[i] = true;
+      }
     }
   }
 
   return Status::OK();
 }
 
-Status CoreWorkerObjectInterface::Wait(const std::vector<ObjectID> &object_ids,
+Status CoreWorkerObjectInterface::Wait(StoreProviderType type,
+                                       const std::unordered_set<ObjectID> &object_ids,
                                        int num_objects, int64_t timeout_ms,
-                                       std::vector<bool> *results) {
-  return store_providers_[StoreProviderType::PLASMA]->Wait(
-      object_ids, num_objects, timeout_ms, worker_context_.GetCurrentTaskID(), results);
+                                       std::unordered_set<ObjectID> *results) {
+  std::vector<ObjectID> ids(object_ids.begin(), object_ids.end());
+  if (!ids.empty()) {
+    std::vector<bool> objects;
+    RAY_RETURN_NOT_OK(store_providers_[type]->Wait(
+        ids, num_objects, timeout_ms, worker_context_.GetCurrentTaskID(), &objects));
+    RAY_CHECK(ids.size() == objects.size());
+    for (size_t i = 0; i < objects.size(); i++) {
+      if (objects[i]) {
+        (*results).insert(ids[i]);
+      }
+    }
+  }
+
+  return Status::OK();
 }
 
 Status CoreWorkerObjectInterface::Delete(const std::vector<ObjectID> &object_ids,
                                          bool local_only, bool delete_creating_tasks) {
-  return store_providers_[StoreProviderType::PLASMA]->Delete(object_ids, local_only,
-                                                             delete_creating_tasks);
+  // There can be a few cases here:
+  // - for task return objects, find the store provider type for an object from
+  //   its transport, and then try to get from the corresponding store provider;
+  // - for other objects, try to get from plasma store.
+  EnumUnorderedMap<StoreProviderType, std::unordered_set<ObjectID>>
+      object_ids_per_store_provider;
+  GetObjectIdsPerStoreProvider(object_ids, &object_ids_per_store_provider);
+
+  for (const auto &entry : object_ids_per_store_provider) {
+    auto type = entry.first;
+    bool is_plasma = (type == StoreProviderType::PLASMA);
+
+    std::vector<ObjectID> ids(entry.second.begin(), entry.second.end());
+    RAY_RETURN_NOT_OK(store_providers_[type]->Delete(
```

Comment:
Switching on `is_plasma` for the boolean flags here is really ugly when we already have separate implementations for different store providers. Why is this necessary? Can't it be handled inside of the store providers?

Reply:
this is because other store provider doesn't support deleting tasks or cross-machine objects, this is really specific to plasma. Another option is to remove the check here and ignore the unsupported flags in other store providers. I'm ok in both ways, kindly let me know your preference.